README.md
# THREE AR

## Introduction

THREE AR aims to provide a simple to use Augmented Reality API for the web, leveraging three.js's 3D capabilities. THREE AR is written in [TypeScript](https://www.typescriptlang.org/). You can see a description of the aims of THREE AR [here](./AIMS.md).

## Usage

THREE AR works with npm and also as a CDN. For npm you can install in the following manner:

```
npm install threear
```

using a CDN you can include a script tag from unpkg like so:

```html
<script src="https://unpkg.com/threear"></script>
```

## Demo 

The examples folder features [a basic demo](https://JamesLMilner.github.io/THREEAR/examples/basic.html), you can open it on a mobile device and point your camera at the following marker:

<p align="center">
  <img width="400px" alt="Hiro Marker" src="./data/hiro.jpg"/>
</p>

## Documentation and Examples

* [Docs](https://JamesLMilner.github.io/THREEAR/docs)
* [Examples (Source)](https://JamesLMilner.github.io/THREEAR/)
* [Examples (Live)](https://JamesLMilner.github.io/THREEAR/)
* [Change Log](./CHANGELOG.md)

## Development and Contribution

Please see the [development guide](./DEVELOPMENT.md). If you are interested in contributing, it may be a good starting point to see the list of open issues on our [GitHub issues page](https://github.com/JamesLMilner/THREEAR/issues). Please take a moment to read the [code of conduct](./CODE_OF_CONDUCT.md).


## Acknowledgements

 - [ar.js](https://github.com/jeromeetienne/AR.js) on which this library is based
 - [jsartoolkit](https://github.com/artoolkitx/jsartoolkit5) the Emscripten library from the C code for artoolkit, on which this library is underpinned
 

## License

All files within the `src/artoolkit` folder are licensed LGPLv3

All other files are MIT Licensed and are adapted from the [work of Jerome Etienne](https://github.com/jeromeetienne/AR.js/blob/master/LICENSE.txt) 


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [v1.0.0-alpha.5](https://github.com/JamesLMilner/THREEAR/compare/v1.0.0-alpha.4...v1.0.0-alpha.5)

> 29 July 2019

- auto-changelog [`#92`](https://github.com/JamesLMilner/THREEAR/pull/92)
- Move checking for lost markers out of marker detection, add example [`#91`](https://github.com/JamesLMilner/THREEAR/pull/91)
- Security updates [`#88`](https://github.com/JamesLMilner/THREEAR/pull/88)
- Fix link to changelog [`#82`](https://github.com/JamesLMilner/THREEAR/pull/82)
- Add the changelog to the readme [`#81`](https://github.com/JamesLMilner/THREEAR/pull/81)
- Update change log for alpha.4 [`#80`](https://github.com/JamesLMilner/THREEAR/pull/80)

#### [v1.0.0-alpha.4](https://github.com/JamesLMilner/THREEAR/compare/v1.0.0-alpha.3...v1.0.0-alpha.4)

> 12 May 2019

- v1.0.0-alpha.4 [`#79`](https://github.com/JamesLMilner/THREEAR/pull/79)
- Use smaller build of artoolkit js [`#78`](https://github.com/JamesLMilner/THREEAR/pull/78)
- Fix the webpack build to account for module types in THREE externals [`#77`](https://github.com/JamesLMilner/THREEAR/pull/77)
- Fix z-fighting bug [`#71`](https://github.com/JamesLMilner/THREEAR/pull/71)
- Added Hole in the Floor example [`#74`](https://github.com/JamesLMilner/THREEAR/pull/74)

#### [v1.0.0-alpha.3](https://github.com/JamesLMilner/THREEAR/compare/v1.0.0-alpha.2...v1.0.0-alpha.3)

> 24 April 2019

- 1.0.0-alpha.3 [`#73`](https://github.com/JamesLMilner/THREEAR/pull/73)
- Added snake example/showcase [`#70`](https://github.com/JamesLMilner/THREEAR/pull/70)
- Fix create/dispose behaviour and example [`#72`](https://github.com/JamesLMilner/THREEAR/pull/72)
- Fixed missing types, made paramaters use Partial [`#66`](https://github.com/JamesLMilner/THREEAR/pull/66)
- Add constructor parameter type of marker classes [`#69`](https://github.com/JamesLMilner/THREEAR/pull/69)
- Exclude Three.js from THREEAR libary [`#68`](https://github.com/JamesLMilner/THREEAR/pull/68)
- Reject properly in Controller on initialisation failure [`#60`](https://github.com/JamesLMilner/THREEAR/pull/60)
- Added `types` entry to package.json [`#58`](https://github.com/JamesLMilner/THREEAR/pull/58)
- Refactor Controller logic for tracking markers, add markerLost event [`#57`](https://github.com/JamesLMilner/THREEAR/pull/57)
- Add package.json meta data [`#55`](https://github.com/JamesLMilner/THREEAR/pull/55)
- Add object shadows example [`#51`](https://github.com/JamesLMilner/THREEAR/pull/51)
- Add shadow example [`7d0487d`](https://github.com/JamesLMilner/THREEAR/commit/7d0487d108d8d21f0e49694c5666b9b2b8fbfe9d)

#### [v1.0.0-alpha.2](https://github.com/JamesLMilner/THREEAR/compare/v1.0.0-alpha.1...v1.0.0-alpha.2)

> 17 March 2019

- V1.0.0 alpha.2 release [`#49`](https://github.com/JamesLMilner/THREEAR/pull/49)
- Add dispose lifecycle methods and example of how to use them [`#48`](https://github.com/JamesLMilner/THREEAR/pull/48)
- Fix minConfidence issue for PatternMarker [`#47`](https://github.com/JamesLMilner/THREEAR/pull/47)
- Refactor Source, add example of custom parent [`#45`](https://github.com/JamesLMilner/THREEAR/pull/45)
- Add facingMode parameter for webcam Source [`#44`](https://github.com/JamesLMilner/THREEAR/pull/44)
- Add basic documentation using TypeDoc [`#39`](https://github.com/JamesLMilner/THREEAR/pull/39)
- Remove runtime allocation of Module properties and functions [`#38`](https://github.com/JamesLMilner/THREEAR/pull/38)
- Center image, new line spacing, minor rewording [`#37`](https://github.com/JamesLMilner/THREEAR/pull/37)
- Move development instructions to seperate markdown file, add CoC [`#36`](https://github.com/JamesLMilner/THREEAR/pull/36)
- Add facing mode to webcam [`b49ea0c`](https://github.com/JamesLMilner/THREEAR/commit/b49ea0c1f296aa9e9fb924d98edeb201d75180ef)
- Fix minConfidence issue for PatternMarker, create BaseMarker abstract class [`517cbf9`](https://github.com/JamesLMilner/THREEAR/commit/517cbf95231bcd4473d0568eb29c03b9c5854b80)
- Move development instructions to seperate markdown file, add Code of Conduct [`5c96302`](https://github.com/JamesLMilner/THREEAR/commit/5c9630224bdc5cee3bdab576a6034e13a5ad80c4)

#### v1.0.0-alpha.1

> 3 March 2019

- Add source maps, .npmignore, change package.json to v1.0.0-alpha.1 [`#35`](https://github.com/JamesLMilner/THREEAR/pull/35)
- Use local vendor folder instead of CDN [`#34`](https://github.com/JamesLMilner/THREEAR/pull/34)
- Inline camera_para.dat [`#33`](https://github.com/JamesLMilner/THREEAR/pull/33)
- Add image and video examples [`#32`](https://github.com/JamesLMilner/THREEAR/pull/32)
-  Prevent initalisation resize  [`#30`](https://github.com/JamesLMilner/THREEAR/pull/30)
- Remove leftover files, use tabs, add prettier to lint command [`#29`](https://github.com/JamesLMilner/THREEAR/pull/29)
- Add prettier webpack plugin and pretty the things [`#27`](https://github.com/JamesLMilner/THREEAR/pull/27)
- Massively simplify API [`#28`](https://github.com/JamesLMilner/THREEAR/pull/28)
- Restructure folders, rename APIs, add artoolkit folder/license [`#25`](https://github.com/JamesLMilner/THREEAR/pull/25)
- Add stats example, vendor script dependencies [`#24`](https://github.com/JamesLMilner/THREEAR/pull/24)
- Add marker positioning smoothing [`#21`](https://github.com/JamesLMilner/THREEAR/pull/21)
- Add strict TypeScript checks [`#15`](https://github.com/JamesLMilner/THREEAR/pull/15)
- Add initial travis.yml file [`#14`](https://github.com/JamesLMilner/THREEAR/pull/14)
- Remove references to tango and aruco [`#13`](https://github.com/JamesLMilner/THREEAR/pull/13)
- Add ARSmoothedControls, fix orientation problem [`#11`](https://github.com/JamesLMilner/THREEAR/pull/11)

src/Profile.ts
/**
 * ArToolkitProfile helps you build parameters for artoolkit
 * - it is fully independent of the rest of the code
 * - all the other classes are still expecting normal parameters
 * - you can use this class to understand how to tune your specific usecase
 * - it is made to help people to build parameters without understanding all the underlying details.
 */
export class Profile {
    private sourceParameters: any;
    private contextParameters: any;
    private defaultMarkerParameters: any;
    private parameters: any;

    constructor(parameters: any) {
        this.reset(parameters.baseURL);
        this.performance("default");
    }

    /**
     * reset all parameters
     */
    public reset(baseURL: string) {
        this.sourceParameters = {
            // to read from the webcam
            sourceType: "webcam"
        };

        this.contextParameters = {
            cameraParametersUrl: baseURL + "../data/camera_para.dat",
            detectionMode: "mono"
        };

        this.defaultMarkerParameters = {
            type: "pattern",
            patternUrl: baseURL + "../data/patt.hiro",
            changeMatrixMode: "modelViewMatrix"
        };

        return this;
    }

    public performance(label: string) {
        if (label === "default") {
            label = this._guessPerformanceLabel();
        }

        if (label === "desktop-fast") {
            this.contextParameters.canvasWidth = 640 * 3;
            this.contextParameters.canvasHeight = 480 * 3;

            this.contextParameters.maxDetectionRate = 30;
        } else if (label === "desktop-normal") {
            this.contextParameters.canvasWidth = 640;
            this.contextParameters.canvasHeight = 480;

            this.contextParameters.maxDetectionRate = 60;
        } else if (label === "phone-normal") {
            this.contextParameters.canvasWidth = 80 * 4;
            this.contextParameters.canvasHeight = 60 * 4;

            this.contextParameters.maxDetectionRate = 30;
        } else if (label === "phone-slow") {
            this.contextParameters.canvasWidth = 80 * 3;
            this.contextParameters.canvasHeight = 60 * 3;

            this.contextParameters.maxDetectionRate = 30;
        } else {
            console.assert(false, "unknonwn label " + label);
        }
        return this;
    }

    public defaultMarker() {
        this.contextParameters.detectionMode = "mono";
        this.defaultMarkerParameters.type = "pattern";
        this.defaultMarkerParameters.patternUrl =
            this.parameters.baseURL + "../data/patt.hiro";

        return this;
    }

    public sourceWebcam() {
        this.sourceParameters.sourceType = "webcam";
        delete this.sourceParameters.sourceUrl;
        return this;
    }

    public sourceVideo(url: string) {
        this.sourceParameters.sourceType = "video";
        this.sourceParameters.sourceUrl = url;
        return this;
    }

    public sourceImage(url: string) {
        this.sourceParameters.sourceType = "image";
        this.sourceParameters.sourceUrl = url;
        return this;
    }

    public changeMatrixMode(changeMatrixMode: any) {
        this.defaultMarkerParameters.changeMatrixMode = changeMatrixMode;
        return this;
    }

    private _guessPerformanceLabel() {
        const isMobile =
            navigator.userAgent.match(/Android/i) ||
            navigator.userAgent.match(/webOS/i) ||
            navigator.userAgent.match(/iPhone/i) ||
            navigator.userAgent.match(/iPad/i) ||
            navigator.userAgent.match(/iPod/i) ||
            navigator.userAgent.match(/BlackBerry/i) ||
            navigator.userAgent.match(/Windows Phone/i)
                ? true
                : false;
        if (isMobile === true) {
            return "phone-normal";
        }
        return "desktop-normal";
    }
}

export default Profile;
